Diffusion/Common/DiffusionImage.swift

```diff
@@ -0,0 +1,143 @@
+        self.steps = steps
+        self.positivePrompt = positivePrompt
+        self.negativePrompt = negativePrompt
+        self.guidanceScale = guidanceScale
+        self.disableSafety = disableSafety
+        self.scheduler = scheduler
+        // Initially set the fileURL to the top level applicationDirectory to allow running the completed instance func save() where the fileURL will be updated to the correct location.
+        self.fileURL = URL.applicationDirectory
+        // init the instance fully before executing an instance function
+        super.init()
+        if let url = save(cgImage: cgImage, filename: genname) {
+            self.fileURL = url
+        } else {
+            fatalError("Fatal error init of DiffusionImage, cannot create image file at \(genname)")
+        }
+    }
+    
+    func encode(with coder: NSCoder) {
+        coder.encode(id, forKey: "id")
+        coder.encode(seed, forKey: "seed")
+        coder.encode(steps, forKey: "steps")
+        coder.encode(positivePrompt, forKey: "positivePrompt")
+        coder.encode(negativePrompt, forKey: "negativePrompt")
+        coder.encode(guidanceScale, forKey: "guidanceScale")
+        coder.encode(disableSafety, forKey: "disableSafety")
+        coder.encode(scheduler, forKey: "scheduler")
+        // Encode cgImage as data
+        if let data = pngRepresentation() {
+            coder.encode(data, forKey: "cgImage")
+        }
+    }
+    
+    required init?(coder: NSCoder) {
+        guard let id = coder.decodeObject(forKey: "id") as? UUID else {
+            return nil
+        }
+        
+        self.id = id
+        self.seed = UInt32(coder.decodeInt32(forKey: "seed"))
+        self.steps = coder.decodeDouble(forKey: "steps")
+        self.positivePrompt = coder.decodeObject(forKey: "positivePrompt") as? String ?? ""
+        self.negativePrompt = coder.decodeObject(forKey: "negativePrompt") as? String ?? ""
+        self.guidanceScale = coder.decodeDouble(forKey: "guidanceScale")
+        self.disableSafety = coder.decodeBool(forKey: "disableSafety")
+        // Should we error/throw here instead of assuming our default scheduler? -- dolmere
+        self.scheduler = coder.decodeObject(forKey: "scheduler") as? Diffusion_StableDiffusionScheduler ?? Diffusion_StableDiffusionScheduler.dpmSolverMultistepScheduler
+        let genname = "\(seed)-\(positivePrompt)".first200Safe
+        
+        // Decode cgImage from data
+        if let imageData = coder.decodeObject(forKey: "cgImage") as? Data {
+#if os(macOS)
+            if let img = NSBitmapImageRep(data: imageData)?.cgImage {
+                self.cgImage = img
+            } else {
+                fatalError("Fatal error loading data with missing cgImage")
+            }
+#else
+            if let image = UIImage(data: imageData)?.cgImage {
+                self.cgImage = image
+            } else {
+                fatalError("Fatal error loading data with missing cgImage")
+            }
+#endif
```

Comment:
Could we use a cgImageFromData helper here too?

Reply:
The reason this one wasn't abstracted out is that it's inside the init method so calling a function to set a value prior to super.init() being called is a compile time error. 

If not for the compiler error ` 'self' used in method call 'cgImageFromData' before 'super.init' ` call  I could envision :

DiffusionImage.swift
`            let img = cgImageFromData(imageData: imageData)`

DiffusionImage+macOS.swift
  `  func cgImageFromData(imageData: Data) -> CGImage {
        if let image = NSBitmapImageRep(data: imageData)?.cgImage {
            return image
        } else {
            fatalError("Fatal error loading data with missing cgImage")
        }
    }`

DiffusionImage+iOS.swift
`
    func cgImageFromData(imageData: Data) -> CGImage {
        if let image = UIImage(data: imageData)?.cgImage {
            return image
        } else {
            fatalError("Fatal error loading data with missing cgImage")
        }
    }`

Since this won't work I could see either putting the platform test in Utils -or- creating an additional platform specific Utils file.

Utils_iOS:
`extension CGImage {
    static func fromData(_ imageData: Data) -> CGImage? {
        if let image = UIImage(data: imageData)?.cgImage {
            return image
        }
        return nil
    }
}`

Utils_macOS:
`extension CGImage {
    static func fromData(_ imageData: Data) -> CGImage? {
        if let image = NSBitmapImageRep(data: imageData)?.cgImage {
            return image
        }
        return nil
    }
}`

DiffusionImage.swift:
`        // Decode cgImage from data
        if let imageData = coder.decodeObject(forKey: "cgImage") as? Data {
            guard let img = CGImage.fromData(imageData) else { fatalError("Fatal error loading data with missing cgImage in object") }
            self.cgImage = img
        } else {
            fatalError("Fatal error loading data with missing cgImage in object")
        }
        self.fileURL = URL.applicationDirectory
        super.init()
`